Issue 5769 dev mode varaible

andrewjt71 opened the pull request:
**Issue addressed**
https://github.com/composer/composer/issues/5769

**Summary**
Sets a php environment variable during the composer install/update, to expose to any post-install/update scripts which mode the install/update was run in.

**Example Usage (In post install/update script)**

```
if (getenv("COMPOSER_DEV_MODE") ==1) {
    // do something
}
```

alcohol commented on doc/03-cli.md:
```diff
@@ -94,7 +94,9 @@ resolution.
-  generation skips the `autoload-dev` rules.
+  generation skips the `autoload-dev` rules. This sets a `COMPOSER_DEV_MODE` php
+  environment variable to true, to expose to any post-install commands whether the
+  install was run in dev mode or not. that.
```
"that." ?

andrewjt71 replied:
I have no idea where `that.` came from! Long day!

alcohol commented on doc/03-cli.md:
```diff
@@ -139,7 +141,10 @@ php composer.phar update vendor/*
+* **--no-dev:** Skip installing packages listed in `require-dev`. The autoloader
+  generation skips the `autoload-dev` rules. This sets a `COMPOSER_DEV_MODE` php
+  environment variable to true, to expose to any post-update commands whether the
+  update was run in dev mode or not. that.
```
Same as above.

alcohol commented on src/Composer/Installer.php:
```diff
@@ -291,7 +291,11 @@ public function run()
                 $this->autoloadGenerator->dump($this->config, $localRepo, $this->package, $this->installationManager, 'composer', $this->optimizeAutoloader);
             }
 
+            // set php environment variable so post install/update scripts can determine which mode install/update was run in
```
I don't think this comment is necessary. The code speaks for itself.

alcohol commented on src/Composer/Installer.php:
```diff
@@ -291,7 +291,11 @@ public function run()
+            // set php environment variable so post install/update scripts can determine which mode install/update was run in
             if ($this->runScripts) {
+                $devMode = $this->devMode;
+                putenv("COMPOSER_DEV_MODE=$this->devMode");
```
I would suggest casting it to an integer, 0 or 1 is easiest to deal with in most shell scripts.

andrewjt71 replied:
👍  Good point, will change this

andrewjt71 commented:
Hi @alcohol, thanks for your feedback. I have addressed your points except the one regarding updating documentation in the `docs` directory. I assume you're referring to the `doc` directory, but this is the directory in which the documentation I initially modified existed (`doc/03-cli`). Have removed the docs I created, but happy to add some documentation around this feature if you give me some more guidance on where they should live. Thanks!

alcohol commented:
@andrewjt71 yes sorry, I should have phrased that better. It would be great if you could document this environment variable in the following section: https://github.com/composer/composer/blob/master/doc/03-cli.md#environment-variables

stof commented:
@alcohol it should not be documented in the cli.md chapter (which lists env variables allowing to control composer). It should be documented in the doc about scripts

andrewjt71 commented:
Hi @alcohol, have added some docs as requested. Feel free to provide any additional feedback if required. Thanks!

alcohol commented:
I rephrased it slightly and applied some wordwrapping. Thanks!

andrewjt71 commented:
Thanks for reviewing @alcohol!

Seldaek commented:
Thanks
